### PR TITLE
add how to develop an add-on using add-on framework

### PR DESCRIPTION
**Pull request by zhiweiyin318**
Signed-off-by: Zhiwei Yin <zyin@redhat.com>

task: https://github.com/open-cluster-management-io/OCM/issues/32

add a sidebar Developer Guides like this 
image

### CODE REVIEW

**Comment by zhiweiyin318** [comment]
@qiujian16  is it better  add a handbook or develop guidebook section on the website to include these code develop guide?

File: content/en/developer-guides/addon.md [hunk 1]
```diff
@@ -0,0 +1,633 @@
...
+
+<!-- spellchecker-disable -->
+
+{{< toc >}}
+
+<!-- spellchecker-enable -->
+
+## Overview
+
+Add-on is an extension which can work with multiple clusters based on the foundation components in open-cluster-management.
```
Comment (mikeshng) [inline content/en/developer-guides/addon.md:16]:
```suggestion
Add-on is an extension which can work with multiple clusters based on the foundation components in open-cluster-management. An add-on is needed to allow for different configurations for different managed clusters. An add-on is also needed to read data from the hub cluster.
```

**Comment by mikeshng** [comment]
This guide is excellent and provides a lot of details for addon development. Do you think it will help if the beginning of the portion of the guide be something thats easier to get started? I am thinking of "if you want to create an addon with helm chart (or yaml manifests) then clone this sample github project (or run this make scaffold command) replace existing helm chart in folder with yours, and modify the following...". This way the beginner has a full addon project with Makefile, go.mod, main.go, manifest folder etc all ready to go right from the beginning in one command (git clone or make scaffold)

File: content/en/developer-guides/addon.md [hunk 2]
```diff
@@ -0,0 +1,633 @@
...
+
+<!-- spellchecker-disable -->
+
+{{< toc >}}
+
+<!-- spellchecker-enable -->
+
+## Overview
+
+Add-on is an extension which can work with multiple clusters based on the foundation components in open-cluster-management. An add-on is needed to allow for different configurations for different managed clusters. An add-on is also needed to read data from the hub cluster.
```
Comment (robinbobbitt) [inline content/en/developer-guides/addon.md:16]: I am new to the addon framework and have struggled with the existing documentation to really understand the purpose, value, and high-level concepts. Some very clear examples of add-ons and the value the framework brings over just creating a controller that applies manifestwork would be really helpful to get somebody started.

Would the following edits, or similar, make sense, based on what I think you're saying?
Add-ons are Open Cluster Management-based extensions that can be used to take action on one or more clusters. Add-ons can support different configurations for different managed clusters, and can also be used to read data from the hub cluster. For example, you might use an add-on to .... (add some concrete examples here).

File: content/en/developer-guides/addon.md [hunk 3]
```diff
@@ -0,0 +1,633 @@
...
+A typical add-on should consist of two kinds of components:
+
+**Add-on agent:** A component in the managed clusters that manages the managed clusters for the hub admins.
```
Comment (robinbobbitt) [inline content/en/developer-guides/addon.md:19]: What more can we say here to be more clear? I think an agent is *typically* a container running on the managed cluster? I'm not clear on saying that the agent "manages" the cluster "for the hub admins"... Would it be correct to say that the agent is a container that performs an action on the managed cluster? And then give an example or two...
And if there is no container, is it an agent-less add-on? You talk later in the document about addons with no agent... What is that?

File: content/en/developer-guides/addon.md [hunk 4]
```diff
@@ -0,0 +1,633 @@
...
+
+**Add-on manager:** A kubernetes controller in the hub cluster that applies the add-on agent manifests to the managed clusters
```
Comment (robinbobbitt) [inline content/en/developer-guides/addon.md:21]: Shall we say it generates and applies the manifests?

File: content/en/developer-guides/addon.md [hunk 5]
```diff
@@ -0,0 +1,633 @@
...
+
+**Add-on agent:** A component in the managed clusters that manages the managed clusters for the hub admins.
+
+**Add-on manager:** A kubernetes controller in the hub cluster that applies the add-on agent manifests to the managed clusters 
+via the ManifestWork API. The manager also can optionally manage the lifecycle of add-on.
+
+There are 2 API resources for add-on in the OCM hub cluster:
+
+**ClusterManagementAddOn:** This is a cluster-scoped resource which allows the user to discover which add-on is available 
+for the cluster manager and also provides metadata information about the add-on.
```
Comment (robinbobbitt) [inline content/en/developer-guides/addon.md:27]: metadata information such as...?

File: content/en/developer-guides/addon.md [hunk 6]
```diff
@@ -0,0 +1,633 @@
...
+**Add-on manager:** A kubernetes controller in the hub cluster that applies the add-on agent manifests to the managed clusters 
+via the ManifestWork API. The manager also can optionally manage the lifecycle of add-on.
+
+There are 2 API resources for add-on in the OCM hub cluster:
+
+**ClusterManagementAddOn:** This is a cluster-scoped resource which allows the user to discover which add-on is available 
+for the cluster manager and also provides metadata information about the add-on. 
+The name of the `ClusterManagementAddOn` resource will be used for the namespace-scoped `ManagedClusterAddOn` resource.
+
+**ManagedClusterAddOn:** This is a namespace-scoped resource which is used by add-on managers to convey their states.
```
Comment (robinbobbitt) [inline content/en/developer-guides/addon.md:30]: Isn't the primary use case of the ManagedClusterAddOn to actually trigger the add-on to be installed on the managed cluster? Shall we call that out first?

File: content/en/developer-guides/addon.md [hunk 7]
```diff
@@ -0,0 +1,633 @@
...
+
+In the busybox example above, we found the `AVAILABLE` status of the `ManagedClusterAddOn` is always `Unknown`.
+
+That’s because the add-on manager did not monitor the status of the add-on agent from the hub cluster.
+
+We support 3 kinds of health prober types to monitor the healthiness of add-on agents.
+
+1. **Lease**
+
+    The add-on agent maintains a `Lease` in its installation namespace with its status, the `registration agent` will
```
Comment (robinbobbitt) [inline content/en/developer-guides/addon.md:270]: What exactly is meant by a Lease? Can we be more clear?

File: content/en/developer-guides/addon.md [hunk 8]
```diff
@@ -0,0 +1,633 @@
...
+    leaseUpdater := lease.NewLeaseUpdater(spokeKubeClient, addonName, installNamespace)
+    go leaseUpdater.Start(context.Background())
+    ```
+
+    `Lease` is the default prober type for add-on, there is nothing to configure for the add-on manager.
+
+2. **Work**
+
+    `Work` health prober indicates the healthiness of the add-on is equal to the overall dispatching status of the 
+    corresponding the ManifestWork resources. It's applicable to those add-ons that don't have a local agent instance
```
Comment (robinbobbitt) [inline content/en/developer-guides/addon.md:285]: Here we're implying some add-ons don't have an agent. What does that mean? Should we talk about this use case at the top of the document? I think we're saying sometimes there's no container.. Or perhaps there's a container but it's not updating the lease...

File: content/en/developer-guides/addon.md (hunk 7 above)
Comment (robinbobbitt) [reply content/en/developer-guides/addon.md:270]: What is the registration agent? Should we introduce this somewhere?

File: content/en/developer-guides/addon.md (hunk 2 above)
Comment (robinbobbitt) [reply content/en/developer-guides/addon.md:16]: If there's not a good example of "different configurations for different managed clusters" below, we should definitely add one!

**Review by robinbobbitt — COMMENTED** [review]
This is a really nice guide that with some additional explanation up front will go a long way to get a developer going with the addon framework. I've added some comments based on my own experiences trying to understand the addon framework.

File: content/en/developer-guides/addon.md (hunk 3 above)
Comment (zhiweiyin318) [reply content/en/developer-guides/addon.md:19]: we also support a health-prober interface which user can use to customize the health check. https://github.com/open-cluster-management-io/addon-framework/blob/bd33a478fe920f268b341afd9fd8d34c446bbf31/pkg/addonfactory/addonfactory.go#L93

File: content/en/developer-guides/addon.md (hunk 2 above)
Comment (zhiweiyin318) [reply content/en/developer-guides/addon.md:16]: thanks for your great suggestions and comments, we discussed this in the community meeting, and keep working on this to support some better examples.

File: content/zh/developer-guides/addon.md [hunk 9]
```diff
@@ -0,0 +1,633 @@
...
+
+We have some available add-ons in the OCM community:
+
+* [cluster-proxy](https://github.com/open-cluster-management-io/cluster-proxy)
+* [managed-serviceaccount](https://github.com/open-cluster-management-io/managed-serviceaccount)
+* [application-manager](https://github.com/open-cluster-management-io/multicloud-operators-subscription)
+* [config-policy-controller](https://github.com/open-cluster-management-io/governance-policy-addon-controller)
+* [governance-policy-framework](https://github.com/open-cluster-management-io/governance-policy-addon-controller)
+
+## Write your first add-on
```
[the remark this answers is not in the document]
Comment (zhiweiyin318) [reply content/zh/developer-guides/addon.md:51]: added the busybox example in addon-framework repo.

File: content/en/developer-guides/addon.md (hunk 2 above)
Comment (zhiweiyin318) [reply content/en/developer-guides/addon.md]: added examples here.

File: content/en/developer-guides/addon.md (hunk 4 above)
Comment (zhiweiyin318) [reply content/en/developer-guides/addon.md]: fixed.

File: content/en/developer-guides/addon.md (hunk 5 above)
Comment (zhiweiyin318) [reply content/en/developer-guides/addon.md]: done

File: content/en/developer-guides/addon.md (hunk 8 above)
Comment (zhiweiyin318) [reply content/en/developer-guides/addon.md]: changed to `It's applicable to those add-ons that don't have a container agent in the managed clusters or don't expect to add `Lease` for the agent container. `

**Comment by zhiweiyin318** [comment]
I addressed the comments and added hosted mode in the guide. please help review again.
 I think we can add a followup to improve the user experience like to enhance our add-on examples or find a proper way to make it easy to build a new addon project etc.. @mikeshng @qiujian16

**Pull request merged**